src/python/pants/backend/python/rules/repl.py

```diff
@@ -39,28 +49,34 @@ class IPythonRepl(ReplImplementation):
 
 
 @rule
-async def run_ipython_repl(repl: IPythonRepl, ipython: IPython) -> ReplBinary:
-    addresses = Addresses(tgt.address for tgt in repl.targets)
-    two_step_pex = await Get(
-        TwoStepPex,
-        TwoStepPexFromTargetsRequest(
-            PexFromTargetsRequest(
-                addresses=addresses,
-                output_filename="ipython-repl.pex",
-                entry_point=ipython.get_entry_point(),
-                additional_requirements=ipython.get_requirement_specs(),
-            )
+async def create_ipython_repl_request(repl: IPythonRepl, ipython: IPython) -> ReplRequest:
+    pex_request = Get(
+        Pex,
+        PexFromTargetsRequest(
+            (tgt.address for tgt in repl.targets),
+            output_filename="ipython.pex",
+            entry_point=ipython.get_entry_point(),
+            additional_requirements=ipython.get_requirement_specs(),
+            include_source_files=True,
         ),
     )
-    repl_pex = two_step_pex.pex
-    return ReplBinary(digest=repl_pex.digest, binary_name=repl_pex.output_filename)
+    source_files_request = Get(
+        UnstrippedPythonSources, UnstrippedPythonSourcesRequest(repl.targets)
+    )
+    pex, source_files = await MultiGet(pex_request, source_files_request)
+    merged_digest = await Get(Digest, MergeDigests((pex.digest, source_files.snapshot.digest)))
+    return ReplRequest(
+        digest=merged_digest,
+        binary_name=pex.output_filename,
+        env={"PEX_EXTRA_SYS_PATH": ":".join(source_files.source_roots)},
+    )
```

Comment:
This is duplicated from above. I tried factoring up, but it made things too complex because of needing to add a new request type.